File: src/main/java/Lec13/Lec13Main.java
package Lec13;

public class Lec13Main {

}

/*
 * 자바 중첩 클래스 종류
 * - static을 사용하는 중첩 클래스 : 밖의 클래스를 직접적으로 참조할 수 없는 클래스
 * - static을 사용하지 않는 중첩 클래스
 *      1. 내부 클래스(Inner Class) : 밖의 클래스를 직접 참조 가능한 클래스
 *      2. 지역 클래스(Local Class) : 메서드 내부에 클래스를 정의 (사용 케이스 거의 없음)
 *      3. 익명 클래스(Anonymous Class) : 일회성 클래스
 *
 * 보통 중첩 클래스라 하면 static 클래스 or 내부 클래스를 말한다.
 */
class JavaHouse {

    private String address;
    private LivingRoom livingRoom;

    public JavaHouse(String address) {
        this.address = address;
        this.livingRoom = new LivingRoom(10);
    }

    public LivingRoom getLivingRoom() {
        return livingRoom;
    }

    /*
     * 내부 클래스
     */
    public class LivingRoom {

        private double area;

        public LivingRoom(double area) {
            this.area = area;
        }

        public String getAddress() {
            return JavaHouse.this.address; // 바깥 클래스와 연결되어 있다.
        }

    }

    /*
     * static 클래스
     */
    public static class StaticLivingRoom {

        private double area;

        public StaticLivingRoom(double area) {
            this.area = area;
        }

        public String getAddress() {
            return JavaHouse.this.address; // static클래스는 바깥 클래스를 바로 불러올 수 없다.
        }

    }

}

/*
 * Effective Java 3rd Edition - Item24, Item86
 *
 * 1. 내부 클래스는 숨겨진 외부 클래스 정보를 가지고 있어,
 * 참조를 해지하지 못하는 경우 메모리 누수가 생길 수 있고,
 * 이를 디버깅 하기 어렵다.
 *
 * 2. 내부 클래스의 직렬화 형태가 명확하게 정의되지 않아
 * 직렬화에 있어 제한이 있다.
 *
 * 즉, 내부 클래스가 외부 클래스를 참조함으로 인에 생기는 몇 가지 문제점이 존재한다.
 *
 * 그래서 Effective Java에서는 "클래스 안에 클래스를 만들 때는 static 클래스를 사용하라"고 가이드하고 있다.
 */

File: src/main/kotlin/Lec13/Lec13Main.kt
package Lec13

/*
 * 코틀린에서 중첩 클래스를 다루는 방법
 *
 * 1. 중첩 클래스의 종류(Lec13Main.java에서 설명)
 * 2. 코틀린의 중첩 클래스와 내부 클래스
 */

/*
 * 2. 코틀린의 중첩 클래스와 내부 클래스
 *
 * Effective Java에서는 "클래스 안에 클래스를 만들 때는 static 클래스를 사용하라"고 가이드하고 있다.
 * 코틀린에서는 위 가이드를 충실하게 따르고 있다.
 *
 * 코틀린에서는 기본적으로 중첩 클래스가 바깥 클래스를 참조하지 않는다.
 * 바깥 클래스를 참조하고 싶다면 inner 키워드를 추가해야 한다.
 */
class KotlinHouse(
    private val address: String,
    private val livingRoom: LivingRoom
) {

    /*
     * 자바에서 권장되는 중첩 클래스 (static class) 작성
     *
     * 코틀린에서 중첩 클래스는 기본적으로 바깥 클래스에 대한 연결이 없는 중첩 클래스가 만들어진다.
     */
    class LivingRoom(
        private val area: Double
    )

    /*
     * 자바에서 권장되지 않는 중첩 클래스 (inner class) 작성
     *
     * 코틀린에서 명시적으로 inner 라는 키워드를 작성해야 한다.
     * 그리고 inner class 내부에서 바깥 클래스에 대한 참조를 위해서는 "this@바깥클래스명" 을 사용한다.
     */
    inner class LivingRoom2(
        private val area: Double
    ) {

        val address: String
            get() = this@KotlinHouse.address

    }

}